core/src/main/scala/kafka/admin/AdminUtils.scala

```diff
@@ -280,30 +280,47 @@ object AdminUtils extends Logging with AdminUtilities {
           s"Assignment: $existingAssignment"))
 
     val partitionsToAdd = numPartitions - existingAssignment.size
-    if (partitionsToAdd <= 0)
+    if (partitionsToAdd < 0)
       throw new InvalidPartitionsException(
         s"The number of partitions for a topic can only be increased. " +
           s"Topic $topic currently has ${existingAssignment.size} partitions, " +
           s"$numPartitions would not be an increase.")
+    else if (partitionsToAdd == 0) {
+      if (replicaAssignment.isDefined) {
+        // Check the provided assignments are compatible with the tail of the existing
+        val newAssignment = replicaAssignment.get
+        // keyed according to the old assignment
+        val newAssignmentWithOldPartitionIds = newAssignment.toList.sortBy { case (p, b) => p }.map { case (p, b) => p - existingAssignment.size + 1 -> b }.toMap
+        val tailOldPartitionIds = existingAssignment.toList.sortBy { case (p, b) => p }.takeRight(newAssignment.size).toMap
+        if (tailOldPartitionIds != newAssignmentWithOldPartitionIds) {
+          val mismatch = tailOldPartitionIds.filter { case (p, b) => newAssignmentWithOldPartitionIds(p) != b }
+          throw new InvalidReplicaAssignmentException(
+            s"Not changing the number of partitions and the given assignments for partitions " +
+              s"${mismatch.keySet.mkString(", ")} are incompatible with the existing assignments " +
+              s"${mismatch.values.map(_.mkString(", ")).mkString("; ")}.")
+        }
```

Comment:
I was thinking about this. Maybe we should leave it as it was as it's consistent with other operations that can be considered idempotent. An example is `createTopics`. Do you agree? If so, sorry for leading you to this path in the first place.

Reply:
I feel conflicted. It *is* confusing for a client if they make a request and get no error, but that the topic is left in a state that's inconsistent with that request. And being consistent with something that is wrong doesn't make this thing right. OTOH, in practice, I suspect that users/clients will explicitly discover the state before making changes, rather than inferring what they think the state should be based on the errors or absence of errors from previous requests. And consistency *is* still important. So I'll remove it. 

Longer term, do you think we should fix those other operations?

Reply:
```I suspect that users/clients will explicitly discover the state before making changes, rather than inferring what they think the state should be based on the errors or absence of errors from previous requests. And consistency is still important.```

This is the reason that I was thinking the additional complexity is probably not worth it. We would have to do the kind of validation you are doing here for every request while users will probably `list/describe` to verify if the operation succeeded since that allows them to potentially check other nodes besides the Controller (which is necessary in many cases).